### PR TITLE
feat(agent-lightning): Integrate agent-lightning system

### PR DESCRIPTION
**Pull request by pppppppppper**
### Description

This PR integrates AgentLightning into the RLinf framework, enabling seamless rollout execution through AgentLightning's store and adapter system. The integration includes:

- **AgentLightningRolloutWorker**: Processes rollout batches using Agent-lightning's LightningStore and TraceToTriplet adapter
- **AgentLightningRLinfRunner**: Training runner that orchestrates the Agent-lightning rollout workflow
- **AgentLightningEvalRunner**: Evaluation runner for checkpoint evaluation
- **SGLangWorkerWithHTTPServer**: Extended SGLang worker with HTTP server to expose OpenAI-compatible API for Agent-lightning integration
- **RLinf Algorithm Interface**: Implements Agent-lightning's `Algorithm` interface, allowing RLinf to be used as a pluggable algorithm within Agent-lightning's framework

### How has this been tested?

train
eval

### Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] Breaking change

### Checklist:
- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: rlinf/runners/agentlightning_runner.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+                    self.cfg.actor.megatron.ckpt_convertor.hf_model_path,
+                    self.cfg.actor.megatron.ckpt_convertor,
+                )
+
+        rollout_handle.wait()
+        if self.use_pre_process_policy:
+            self.rollout.offload_engine().wait()
+
+        server_addresses = []
+        if hasattr(self.rollout, 'http_server_start') and ray is not None:
```
Comment (qurakchin) [inline rlinf/runners/agentlightning_runner.py]: move this process into rollout.init_workers.

File: rlinf/runners/agentlightning_runner.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            for rank in range(num_workers):
+                port = base_port + rank
+                server_addresses.append(f"{node_ip}:{port}")
+            
+            for rank in range(num_workers):
+                self.rollout.execute_on(rank).http_server_start().wait()
+            
+            logging.info(f"[AgentLightningRLinfRunner] Started HTTP servers on {num_workers} workers with addresses {server_addresses}")
+
+        is_eval_mode = False
```
Comment (qurakchin) [inline rlinf/runners/agentlightning_runner.py]: delete this and call the function with a "is_eval_mode=False" param.

File: rlinf/runners/agentlightning_runner.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        self.agentlightning_rollout_worker.init_worker(
```
Comment (qurakchin) [inline rlinf/runners/agentlightning_runner.py:130]: Can this init process parallel with the rollout.init_worker?

File: rlinf/runners/agentlightning_runner.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+                    with self.timer("prepare_data"):
+                        self._put_batch(batch)
+
+                    rollout_handle: Handle = self.agentlightning_rollout_worker.process_rollout_batch(
+                        input_channel=self.dataloader_channel,
+                        output_channel=self.rollout_channel,
+                    )
+
+                    reward_handle = None
```
Comment (qurakchin) [inline rlinf/runners/agentlightning_runner.py]: It looks like that you don't use the reward worker. Do all agent-lightning cases get reward within the agent-lightning framework?

File: rlinf/workers/rollout/sglang/sglang_worker_server.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        await super().init_worker()
+        
+        if self._enable_http_server:
+            self._init_openai_serving()
+            if start_http_server:
+                self.http_server_start()
+
+    def shutdown(self):
+        if self._enable_http_server and self._http_server_task is not None:
+            self._http_server.should_exit = True
```
Comment (qurakchin) [inline rlinf/workers/rollout/sglang/sglang_worker_server.py]: Do you wait until the _http_server_task exit?

File: rlinf/runners/agentlightning_runner.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+            inference,
+            actor,
+            reward,
+        )
+        
+        self.store = store
+        self.adapter = adapter
+        self.agentlightning_rollout_worker = agentlightning_rollout_worker
+
+    def _build_dataloader(self, train_dataset, val_dataset, collate_fn=None):
```
Comment (qurakchin) [inline rlinf/runners/agentlightning_runner.py:63]: What's the difference betweent it and ReasoningRunner._build_dataloader

File: examples/agentlightning/entrypoint/entrypoint.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        )
+    
+    actor_worker_cls = get_actor_worker(cfg)
+    actor_placement_strategy = component_placement.get_strategy("actor")
+    actor_group = actor_worker_cls.create_group(cfg, component_placement).launch(
+        cluster, name=cfg.actor.group_name, placement_strategy=actor_placement_strategy
+    )
+    
+    reward_placement_strategy = component_placement.get_strategy("reward")
+    reward_group = RewardWorker.create_group(cfg, component_placement).launch(
```
Comment (qurakchin) [inline examples/agentlightning/entrypoint/entrypoint.py]: It seems that you don't need a reward worker.

File: rlinf/runners/agentlightning_runner.py (hunk 4 above)
Comment (pppppppppper) [reply rlinf/runners/agentlightning_runner.py]: Rewards are currently handled by the frontend agent, which emits them into the AgentLightning store using `agl.emit_reward()`.

File: rlinf/runners/agentlightning_runner.py (hunk 6 above)
Comment (pppppppppper) [reply rlinf/runners/agentlightning_runner.py:63]: AgentLightning takes raw sample dicts as rollout inputs, so we only collate by grouping fields into lists and pass the raw batch through the channel.

File: examples/agentlightning/calc_x/config/qwen2.5-1.5b-trajectory.yaml [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+  val_rollout_batch_size: null
+  num_workers: 8
+  prompt_key: question
+  answer_key: result
+  apply_chat_template: False
+  shuffle: True
+  validation_shuffle: False
+  seed: 42
+  train_data_paths: ["/mnt/public/yule/data/calc_x/train.parquet"]
+  val_data_paths: ["/mnt/public/yule/data/calc_x/test.parquet"]
```
Comment (thereAreDemonsNearby) [inline examples/agentlightning/calc_x/config/qwen2.5-1.5b-trajectory.yaml]: remove your personal info

File: examples/agentlightning/calc_x/calc_agent.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            result = await asyncio.wait_for(calc_agent.run(task=prompt), timeout=300.0)
+            # evaluate
+            last_message = cast(str, result.messages[-1].content)  # type: ignore
+            answer = re.search(r"###\s*ANSWER:\s*(.+?)(\s*###|$)", last_message)
+            if answer:
+                answer = answer.group(1)
+            else:
+                answer = last_message
+        except asyncio.TimeoutError as e:
+            print("Timeout occurred. Error:", str(e))
```
Comment (thereAreDemonsNearby) [inline examples/agentlightning/calc_x/calc_agent.py]: should we use logging instead of print here? @qurakchin

File: examples/agentlightning/calc_x/calc_agent.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            answer = "None"
+        except Exception as e:
+            print("Failure:", str(e))
+            answer = "None"
+        reward = await evaluate(answer, str(task["result"]))
+        agl.emit_reward(reward)  # Emit reward for tracing
+        # Disabled verbose logging
+        print("answer: {} ground_truth: {} reward: {}".format(answer, task["result"], reward))
```
Comment (thereAreDemonsNearby) [inline examples/agentlightning/calc_x/calc_agent.py]: remove this log

File: rlinf/runners/agentlightning_eval_runner.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        self.init_workers()
+        self._sync_weights()
+        
+        step = 0
+        ckpt_dir = os.path.dirname(actor_checkpoint_path)
+        if os.path.basename(ckpt_dir).startswith("global_step_"):
+            step = int(os.path.basename(ckpt_dir).split("_")[-1])
+        
+        avg_reward = self._run_eval_loop()
+        print(f"Evaluation Results:")
```
Comment (thereAreDemonsNearby) [inline rlinf/runners/agentlightning_eval_runner.py]: should we use logging instead of print here?

File: rlinf/runners/agentlightning_runner.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+                        self.global_steps,
+                        self.max_steps,
+                        self.cfg.runner.val_check_interval,
+                        self.cfg.runner.save_interval,
+                        1.0,
+                        run_time_exceeded=run_time_exceeded,
+                    )
+
+                    if save_model:
+                            self._save_checkpoint()
```
Comment (thereAreDemonsNearby) [inline rlinf/runners/agentlightning_runner.py]: wrong indent

File: examples/agentlightning/calc_x/eval_utils.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def float_eval(input_str: str) -> float:
+    if " = around " in input_str:
+        input_str = input_str.split(" = around ")[0]
+    expr = sympy.parse_expr(input_str, evaluate=True)
+    return float(expr.evalf())
+
+
+def scalar_are_results_same(pred_result: str, true_result: str, rel_tol: float) -> bool:
```
Comment (thereAreDemonsNearby) [inline examples/agentlightning/calc_x/eval_utils.py]: scalar**s** ?

File: examples/agentlightning/entrypoint/entrypoint.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+            cfg=cfg,
+            placement=component_placement,
+            val_dataset=val_dataset,
+            rollout=rollout_group,
+            actor=actor_group,
+            store=store,
+            adapter=adapter,
+            agentlightning_rollout_worker=agentlightning_rollout_group,
+        )
+        runner.eval(checkpoint_dir=eval_checkpoint_dir)
```
Comment (thereAreDemonsNearby) [inline examples/agentlightning/entrypoint/entrypoint.py:97]: does the eval runner need a 'init_workers' call ?

File: rlinf/runners/agentlightning_eval_runner.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+        self.cfg = cfg
+        self.placement = placement
+        self.val_dataset = val_dataset
+        self.rollout = rollout
+        self.actor = actor
+        self.store = store
+        self.adapter = adapter
+        self.agentlightning_rollout_worker = agentlightning_rollout_worker
+
+        from rlinf.scheduler import Channel
```
Comment (thereAreDemonsNearby) [inline rlinf/runners/agentlightning_eval_runner.py]: why not make this import global?

File: rlinf/runners/agentlightning_runner.py (hunk 3 above)
Comment (pppppppppper) [reply rlinf/runners/agentlightning_runner.py:130]: We need the server_addresses produced by rollout.init_worker().wait() and get_server_address() to initialize agentlightning_rollout_worker.init_worker(), so this init step has to run after rollout.init_worker rather than in parallel.

File: examples/agentlightning/entrypoint/entrypoint.py (hunk 14 above)
Comment (pppppppppper) [reply examples/agentlightning/entrypoint/entrypoint.py:97]: runner.eval already calls init_workers() internally (for both original model and actor checkpoint cases), so an extra init_workers call here isn’t needed.
